### ISSUE TITLE
\u0000 is treated as end of string, even if it's not

### ISSUE THREAD

**Issue by greyblake** [sparklemotion/nokogiri#1535]
It's easy to reproduce:

```
2.1.5 :013 > Nokogiri.parse "<a>ABC\u0000XYZ</a>"
 => #<Nokogiri::XML::Document:0x120b3e4 name="document"
   children=[#<Nokogiri::XML::Element:0x120b010 
     name="a"
     children=[#<Nokogiri::XML::Text:0x120ade0 "ABC">]>]> 
```

Everything that comes after `\u0000` is ignored. `XYZ` does not appear in the parsed output.

Unfortunately one of SOAP API returns us such data :-(

**Comment by flavorjones** [comment]
Hi @greyblake,

Unfortunately this behavior is due to the underlying libxml2 implementation, and it's not something that Nokogiri can change.

For what it's worth, here's a libxml2 mailing list response to this question from 2010:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=574104;msg=9

TL;DR:

> For a starter, libxml2 treats your data as UTF-8, and as such uses null
terminated strings, so this is not an unexpected behaviour.
>
> Secondly, the null character is not allowed in a xml file.

Hope this helps,

-m

**Comment by greyblake** [comment]
Ah.. Thank you for the explanation!)
